File: app/src/main/java/forpdateam/ru/forpda/presentation/forum/ForumPresenter.java
package forpdateam.ru.forpda.presentation.forum;

import android.os.Bundle;

import com.arellomobile.mvp.InjectViewState;

import forpdateam.ru.forpda.api.forum.models.ForumItemTree;
import forpdateam.ru.forpda.common.Utils;
import forpdateam.ru.forpda.common.mvp.BasePresenter;
import forpdateam.ru.forpda.model.repository.ForumRepository;
import forpdateam.ru.forpda.ui.TabManager;
import forpdateam.ru.forpda.ui.fragments.TabFragment;
import forpdateam.ru.forpda.ui.fragments.search.SearchFragment;
import forpdateam.ru.forpda.ui.fragments.topics.TopicsFragment;
import io.reactivex.disposables.Disposable;

/**
 * Created by radiationx on 03.01.18.
 */

@InjectViewState
public class ForumPresenter extends BasePresenter<ForumView> {

    private ForumRepository forumRepository;

    public ForumPresenter(ForumRepository forumRepository) {
        this.forumRepository = forumRepository;
    }

    public void loadForums() {
        Disposable disposable
                = forumRepository.getForums()
                .doOnTerminate(() -> getViewState().setRefreshing(true))
                .doAfterTerminate(() -> getViewState().setRefreshing(false))
                .subscribe(forums -> {
                    getViewState().showForums(forums);
                    saveCacheForums(forums);
                }, this::handleErrorRx);
        addToDisposable(disposable);
    }

    public void getCacheForums() {
        Disposable disposable
                = forumRepository.getCache()
                .doOnTerminate(() -> getViewState().setRefreshing(true))
                .doAfterTerminate(() -> getViewState().setRefreshing(false))
                .subscribe(forums -> {
                    if (forums.getForums() == null) {
                        loadForums();
                    } else {
                        getViewState().showForums(forums);
                    }
                }, this::handleErrorRx);
        addToDisposable(disposable);
    }

    private void saveCacheForums(ForumItemTree rootForum) {
        Disposable disposable
                = forumRepository.saveCache(rootForum)
                .doOnTerminate(() -> getViewState().setRefreshing(true))
                .doAfterTerminate(() -> getViewState().setRefreshing(false))
                .subscribe(() -> {

                }, this::handleErrorRx);
        addToDisposable(disposable);
    }

    public void copyLink(ForumItemTree item) {
        Utils.copyToClipBoard("https://4pda.ru/forum/index.php?showforum=" + item.getId());
    }

    public void navigateToForum(ForumItemTree item) {
        Bundle args = new Bundle();
        args.putInt(TopicsFragment.TOPICS_ID_ARG, item.getId());
        TabManager.get().add(TopicsFragment.class, args);
    }

    public void navigateToSearch(ForumItemTree item) {
        String url = "https://4pda.ru/forum/index.php?act=search&source=all&forums%5B%5D=" + item.getId();
        Bundle args = new Bundle();
        args.putString(TabFragment.ARG_TAB, url);
        TabManager.get().add(SearchFragment.class, args);
    }
}

File: app/src/main/java/forpdateam/ru/forpda/presentation/forum/ForumPresenter.kt
package forpdateam.ru.forpda.presentation.forum

import forpdateam.ru.forpda.common.Utils
import forpdateam.ru.forpda.common.mvp.BasePresenter
import forpdateam.ru.forpda.entity.remote.forum.ForumItemFlat
import forpdateam.ru.forpda.extensions.coRunCatching
import forpdateam.ru.forpda.model.data.remote.api.favorites.FavoritesApi
import forpdateam.ru.forpda.model.repository.faviorites.FavoritesRepository
import forpdateam.ru.forpda.model.repository.forum.ForumRepository
import forpdateam.ru.forpda.presentation.ErrorHandler
import forpdateam.ru.forpda.presentation.Screen
import forpdateam.ru.forpda.presentation.TabRouter
import kotlinx.coroutines.launch
import moxy.InjectViewState
import ru.radiationx.coretypes.ForumId
import ru.radiationx.links.Link
import ru.radiationx.quill.QuillExtra

/**
 * Created by radiationx on 03.01.18.
 */

data class ForumExtra(
    val forumId: ForumId?
): QuillExtra

@InjectViewState
class ForumPresenter(
    private val argExtra: ForumExtra,
    private val forumRepository: ForumRepository,
    private val favoritesRepository: FavoritesRepository,
    private val router: TabRouter,
    private val errorHandler: ErrorHandler,
    private val utils: Utils
) : BasePresenter<ForumView>() {

    private var targetForumId: ForumId? = argExtra.forumId

    override fun onFirstViewAttach() {
        super.onFirstViewAttach()
        getCacheForums()
    }

    fun loadForums() {
        viewModelScope.launch {
            viewState.setRefreshing(true)
            coRunCatching {
                forumRepository.getForums()
            }.onSuccess {
                viewState.showForums(it)
                scrollToTarget()
            }.onFailure {
                errorHandler.handle(it)
            }
            viewState.setRefreshing(false)
        }
    }

    private fun getCacheForums() {
        viewModelScope.launch {
            viewState.setRefreshing(true)
            coRunCatching {
                forumRepository.getCache()
            }.onSuccess {
                if (it.isNotEmpty()) {
                    viewState.showForums(it)
                    scrollToTarget()
                }
            }.onFailure {
                errorHandler.handle(it)
            }
            viewState.setRefreshing(false)
            loadForums()
        }
    }

    private fun scrollToTarget() {
        val forumId = targetForumId ?: return
        viewState.scrollToForum(forumId.id)
        targetForumId = null
    }

    fun markRead(id: Int) {
        viewModelScope.launch {
            coRunCatching {
                forumRepository.markRead(id)
            }.onSuccess {
                viewState.onMarkRead()
            }.onFailure {
                errorHandler.handle(it)
            }
        }
    }

    fun markAllRead() {
        viewModelScope.launch {
            coRunCatching {
                forumRepository.markAllRead()
            }.onSuccess {
                viewState.onMarkAllRead()
            }.onFailure {
                errorHandler.handle(it)
            }
        }
    }

    fun addToFavorite(forumId: Int, subType: String) {
        viewModelScope.launch {
            coRunCatching {
                favoritesRepository.editFavorites(
                    FavoritesApi.ACTION_ADD_FORUM,
                    -1,
                    forumId,
                    subType
                )
            }.onSuccess {
                viewState.onAddToFavorite(it)
            }.onFailure {
                errorHandler.handle(it)
            }
        }
    }

    fun copyLink(item: ForumItemFlat) {
        utils.copyToClipBoard("https://4pda.to/forum/index.php?showforum=${item.id}")
    }

    fun navigateToForum(item: ForumItemFlat) {
        router.navigateTo(Screen.Topics(link = Link.Board.Forum.default(ForumId(item.id))))
    }

    fun navigateToSearch(item: ForumItemFlat) {
        val link = Link.Board.Search.default.copy(forums = setOf(Link.Board.Search.Forum.Id(forumId = ForumId(item.id))))
        router.navigateTo(Screen.Search.Forum(link = link))
    }
}
